src/fp.cpp

```diff
@@ -445,133 +514,101 @@ bool fp2::sign() const
+    c.addAssign(e);
     return c;
 }
 
 void fp2::addAssign(const fp2& e)
 {
-    _addAssign(&c0, &e.c0);
-    _addAssign(&c1, &e.c1);
-}
-
-fp2 fp2::ladd(const fp2& e) const
-{
-    fp2 c;
-    _ladd(&c.c0, &c0, &e.c0);
-    _ladd(&c.c1, &c1, &e.c1);
-    return c;
+    _add(&c0, &c0, &e.c0);
+    _add(&c1, &c1, &e.c1);
 }
 
 fp2 fp2::dbl() const
 {
-    fp2 c;
-    _double(&c.c0, &c0);
-    _double(&c.c1, &c1);
+    fp2 c(*this);
+    c.doubleAssign();
     return c;
 }
 
 void fp2::doubleAssign()
 {
-    _doubleAssign(&c0);
-    _doubleAssign(&c1);
+    _double(&c0, &c0);
+    _double(&c1, &c1);
 }
 
-fp2 fp2::ldouble() const
+fp2 fp2::subtract(const fp2& e) const
 {
-    fp2 c;
-    _ldouble(&c.c0, &c0);
-    _ldouble(&c.c1, &c1);
+    fp2 c(*this);
+    c.subtractAssign(e);
     return c;
 }
 
-fp2 fp2::sub(const fp2& e) const
+void fp2::subtractAssign(const fp2& e)
 {
-    fp2 c;
-    _sub(&c.c0, &c0, &e.c0);
-    _sub(&c.c1, &c1, &e.c1);
-    return c;
+    _subtract(&c0, &c0, &e.c0);
+    _subtract(&c1, &c1, &e.c1);
 }
 
-void fp2::subAssign(const fp2& e)
-{
-    _subAssign(&c0, &e.c0);
-    _subAssign(&c1, &e.c1);
-}
-
-fp2 fp2::neg() const
+fp2 fp2::negate() const
 {
     fp2 c;
-    _neg(&c.c0, &c0);
-    _neg(&c.c1, &c1);
+    _negate(&c.c0, &c0);
+    _negate(&c.c1, &c1);
     return c;
 }
 
-fp2 fp2::conj() const
+fp2 fp2::conjugate() const
 {
     fp2 c;
     c.c0 = c0;
-    _neg(&c.c1, &c1);
+    _negate(&c.c1, &c1);
     return c;
 }
 
-fp2 fp2::mul(const fp2& e) const
+fp2 fp2::multiply(const fp2& e) const
 {
-    fp t[4];
-    fp2 c;
-    _mul(&t[1], &c0, &e.c0);
-    _mul(&t[2], &c1, &e.c1);
-    _add(&t[0], &c0, &c1);
-    _add(&t[3], &e.c0, &e.c1);
-    _sub(&c.c0, &t[1], &t[2]);
-    _addAssign(&t[1], &t[2]);
-    _mul(&t[0], &t[0], &t[3]);
-    _sub(&c.c1, &t[0], &t[1]);
+    fp2 c(*this);
+    c.multiplyAssign(e);
     return c;
 }
 
-void fp2::mulAssign(const fp2& e)
+void fp2::multiplyAssign(const fp2& e)
 {
     fp t[4];
-    _mul(&t[1], &c0, &e.c0);
-    _mul(&t[2], &c1, &e.c1);
+    _multiply(&t[1], &c0, &e.c0);
+    _multiply(&t[2], &c1, &e.c1);
     _add(&t[0], &c0, &c1);
     _add(&t[3], &e.c0, &e.c1);
-    _sub(&c0, &t[1], &t[2]);
-    _addAssign(&t[1], &t[2]);
-    _mul(&t[0], &t[0], &t[3]);
-    _sub(&c1, &t[0], &t[1]);
+    _subtract(&c0, &t[1], &t[2]);
+    _add(&t[1], &t[1], &t[2]);
+    _multiply(&t[0], &t[0], &t[3]);
+    _subtract(&c1, &t[0], &t[1]);
 }
 
 fp2 fp2::square() const
 {
-    fp t[3];
-    fp2 c;
-    _ladd(&t[0], &c0, &c1);
-    _sub(&t[1], &c0, &c1);
-    _ldouble(&t[2], &c0);
-    _mul(&c.c0, &t[0], &t[1]);
-    _mul(&c.c1, &t[2], &c1);
+    fp2 c(*this);
+    c.squareAssign();
     return c;
 }
 
 void fp2::squareAssign()
 {
     fp t[3];
     _ladd(&t[0], &c0, &c1);
-    _sub(&t[1], &c0, &c1);
+    _subtract(&t[1], &c0, &c1);
     _ldouble(&t[2], &c0);
-    _mul(&c0, &t[0], &t[1]);
-    _mul(&c1, &t[2], &c1);
+    _multiply(&c0, &t[0], &t[1]);
+    _multiply(&c1, &t[2], &c1);
```

Comment:
Have the above two `_multiply` calls been verified to be safe?

`t[0]` and `[2]` could both be larger than the prime modulus.